### DISCUSSION TITLE
How to enforce state-dependent probability distribution constraints as well as set a specific formula for parameters of state-dependent observation distributions using DM

### DISCUSSION THREAD

**Discussion by Dekike** [bmcclintock/momentuHMM#63] [Q&A] [answered]
Hi Brett,

I have a 2-state ("1. Active/2. Resting") HMM in which, among other things, I want to include the effect of the water temperature on the parameters of the state-dependent distributions of `meanActivity`, which is my data stream. Similarly to what you did in the `Northern fur seal` example, I want to constrain the Gamma parameters such that the "Active" state tends to have higher activity values than "Resting". I guess I have to do this using `DM` and `workBounds` arguments. However, I do not know how to call `DM` such that I include the effect of the water temperature and force meanActivity[state=="Active"] > meanActivity[state=="Resting"].

Similarly to my case, in the `Northern fur seal` example it would be like including water temperature as a covariate in the HMMs to see the effect of water temperature on the parameters of the state-dependent distributions of "step", "angle" and "dive", while keeping the constraints you mentioned. If I am not wrong, in the vignettes there is no example of doing both things at once. Is there a reason?

I show the code I was able to run so far adaptating what you did in the `Northern fur seal` :
```
df <- df[,c("ID","meanActivity","WaterT")]
df <- prepData(data=df,coordNames=NULL,covNames = c("WaterT"))
nbStates <- 2
stateNames <- c("Active","Resting")
dist = list(meanActivity = "gamma")

Par0_m1 <- list(meanActivity=c(1.2,0.15,1.2,0.15))

meanActivityDM<-matrix(c(1,0,0,0,
                 1,1,0,0,
                 0,0,1,0,
                 0,0,0,1),2*length(stateNames),4,byrow=TRUE,
               dimnames=list(c(paste0("mean_",1:length(stateNames)),paste0("sd_",1:length(stateNames))),
                             c("mean_12:(Intercept)","mean_2",
                               paste0("sd_",1:length(stateNames),":(Intercept)"))))

meanActivityworkBounds <- matrix(c(-Inf,-Inf,-Inf,-Inf,Inf,0,rep(Inf,2)),4,2,
                         dimnames=list(colnames(meanActivityDM),c("lower","upper")))

DM<- list(meanActivity=meanActivityDM)

workBounds<- list(meanActivity=meanActivityworkBounds)

Par0 <- getParDM(nbStates = nbStates, dist = dist,                
                 Par = Par0_m1, DM = DM, workBounds = workBounds,    
                 estAngleMean = list(angle = FALSE))

formula <- ~ MWTB

m3 <- fitHMM(data = HJose.InviernoData, nbStates = 2, dist = dist, Par0 = Par0, 
            DM = DM, workBounds = workBounds, stateNames = stateNames, formula = formula)
```
This works. However, if I do not know how to include `WaterT` effect on `DM`. I tried different things:
```
DM<- list(meanActivity=meanActivityDM~WaterT)
DM<- list(meanActivity=list(meanActivityDM=~WaterT))
Etc.
```
What am I doing wrong? If my question is inappropriate I apologize in advance.

**Comment by bmcclintock** [comment, marked as answer]
To include both relational constraints and covariates on state-dependent distribution parameters, one must specify `DM` as a pseudo-design matrix. There are numerous examples using both constraints and covariates on state-dependent distribution parameters throughout the `momentuHMM` [vignette](https://cran.r-project.org/package=momentuHMM/vignettes/momentuHMM.pdf) and [examples](https://github.com/bmcclintock/momentuHMM/tree/master/vignettes/examples). I suggest you take a step back and spend some time carefully going through the entire vignette and all of the examples. See also #42.

**Comment by Dekike** [comment]
Thanks Brett,

I will read more carefully `momentuHMM` vignette and examples.

Keep in touch

**Comment by Dekike** [comment]
Hi Brett,

First of all, thanks for the references. They were very useful to understand how pseudo-design matrices work, especially those indicated in [#42](https://github.com/bmcclintock/momentuHMM/issues/42). I think I got the way for including both relational constraints and covariates on state-dependent distribution parameters using `DM` as a pseudo-design matrix.

However, there is something "rare" that I wanted to share with you. After using `DM` as a pseudo-design matrix, when I use `plot()` I don't see the plot in which estimates (and 95% confidence intervals) for the `Activity` mean parameter is shown as a function of temperature for any of my two states ("Active" and "Resting"). To check if the problem was related with the way I included the relational constraints on state-dependent distribution parameters, I compared "plot()" results for two models, one in which I used "pseudo-design matrix" (`m3`) and another in which I did not (`m2`) but without including any relational constraint. What I saw is that for the model in which I do not use the "pseudo-design matrix" (`m2`), the function "plot()" shows a graph with estimates for the `Activity` mean parameter as a function of temperature for both states. Nevertheless, "plot()" did not show these graphs when I used the "pseudo-design matrix" procedure (`m3`). I share the code I used since I don't know if I made some mistake or if this is done on purpose and I am doing well the "pseudo-design matrix".

```
######  Without using a pseudo-design matrix  #######

stateNames <- c("Active","Resting")
dist = list(meanAcc = "gamma")
nbStates <- 2

Par0_m1 <- list(meanAcc=c(1.1,0.19,1.1,0.1))

m1 <- fitHMM(data = HJoseData, nbStates = 2, dist = dist, Par0 = Par0_m1, stateNames = stateNames)

formula <- ~ Temp
DM <- list(meanAcc = list(mean =~Temp, sd = ~ 1))
Par0_m2 <- getPar0(model=m1, formula=formula, DM=DM)

m2 <- fitHMM(data = HJoseData, nbStates = 2, dist = dist, Par0 = Par0_m2$Par,beta0 = Par0_m2$beta, DM = DM, stateNames = stateNames, formula = formula)

getPar0(m2)
_$Par$meanAcc
mean_1:(Intercept)   mean_1:Temp    mean_2:(Intercept)     mean_2:Temp   sd_1:(Intercept)   sd_2:(Intercept) 
        0.11815810       0.01018682      -1.96815212         0.01958747     -0.43366398      -2.33489454 

$beta
                        1 -> 2           2 -> 1
(Intercept)         -1.55895247       -3.3691830
Temp                 -0.03290427        0.1742653

$delta
[1] 0.5065931 0.4934069_

######   Using a pseudo-design matrix   ######

stateNames <- c("Active","Resting")
dist = list(meanAcc = "gamma")
nbStates <- 2

formula <- ~ Temp

meanAccDM<-matrix(c(1,1,0,0,0,0,
                    0,0,1,1,0,0,
                    0,0,0,0,1,0,
                    0,0,0,0,0,1),2*nbStates,6,byrow=TRUE,                                       
                  dimnames=list(c(paste0("mean_",1:nbStates),paste0("sd_",1:nbStates)),         
                                c("mean_1:(Intercept)","mean_1:Temp","mean_2:(Intercept)","mean_2:Temp", 
                                  paste0("sd_",1:nbStates,":(Intercept)"))))                    

meanAccworkBounds <- matrix(c(rep(-Inf,6),rep(Inf,2),Inf,rep(Inf,3)),6,2, dimnames=list(colnames(meanAccDM), c("lower","upper")))   

DM<-list(meanAcc=meanAccDM)

workBounds<-list(meanAcc=meanAccworkBounds)

Par0 <- list(meanAcc=c(1.1,0.19,1.1,0.1))

Par0 <- getParDM(nbStates = nbStates, dist = dist,
                 Par = Par0, formula=formula, DM = DM, workBounds = workBounds,    
                 estAngleMean = list(angle = FALSE))
Par0  
$meanAcc
[1] -0.3351155  0.5174371 -0.8525526 -0.8525526  0.1823216 -2.0402208

m3 <- fitHMM(data = HJoseData, nbStates = 2, dist = dist, Par0 = Par0, DM = DM, stateNames = stateNames, formula = formula)

getPar0(m3)
$Par
$Par$meanAcc
mean_1:(Intercept)        mean_1:Temp mean_2:(Intercept)        mean_2:Temp   sd_1:(Intercept)   sd_2:(Intercept) 
         0.1504605          0.1504605         -0.8177634         -0.8177634         -0.4301297         -2.3267067 


$beta
                 1 -> 2     2 -> 1
(Intercept) -1.13667276 -3.2930684
Temp        -0.05733824  0.1701075

$delta
[1] 9.999912e-01 8.792998e-06
```
I am able to obtain the plot below using `plot(m2)` but not using `plot(m3)`, although I run both models perfectly without any problem. I have realized that estimates of the `Activity` mean and sd parameters change between `m2` and `m3`, which confuses me a little since I understood that both approaches were equivalent, but I guess this is not the problem.

![Rplot](https://user-images.githubusercontent.com/65897655/95422359-770f1980-093f-11eb-84d2-cf2eb1d0045c.png)

Am I doing something wrong or is this something done on purpose? Is there a way of obtaining the previous graph using your own "Pseudo-design matrix"?

If you want that I post this as a different issue in `GitHub`, just tell me.

**Comment by bmcclintock** [comment]
`Temp` is not being included in the pseudo-design matrix for model `m3` (`meanAccDM`). It should be:

```
meanAccDM<-matrix(c(1,"Temp",0,     0,0,0,
                    0,     0,1,"Temp",0,0,
                    0,     0,0,     0,1,0,
                    0,     0,0,     0,0,1),2*nbStates,6,byrow=TRUE,                                       
                  dimnames=list(c(paste0("mean_",1:nbStates),paste0("sd_",1:nbStates)),         
                                c("mean_1:(Intercept)","mean_1:Temp","mean_2:(Intercept)","mean_2:Temp", 
                                  paste0("sd_",1:nbStates,":(Intercept)"))))
```

**Comment by Dekike** [comment]
Dear Brett,

I missed that small but very important detail.

I really appreciate your time. Thank you very much.

Keep in touch!
